GPT2/GPT2/train.py

```diff
@@ -0,0 +1,169 @@
+import torch
+import torch.nn as nn
+import torch.optim as optim
+from torch.utils.data import Dataset, DataLoader
+from transformers import GPT2Tokenizer, GPT2LMHeadModel
+from transformers import AdamW, get_linear_schedule_with_warmup
+import numpy as np
+import tqdm
+import copy
+from opacus import PrivacyEngine
+
+
+def finetunable_GPT2_params(model, finetune):
+    #works on refactored GPT2
+    def extract_finetune_index(name):
+        #subroutine that parses string 
+        ft_idx = None
+        if 'emb' in name:
+            ft_idx = -1
+        elif name.startswith('transformers'):
+            ft_idx = int(name.split('.')[1])
+        elif 'head' in name:
+            ft_idx = float('inf') #always FT the head
+        return ft_idx
+            
+    params = [] 
+    for name,param in model.named_parameters():
+        if extract_finetune_index(name) >= finetune and param.requires_grad:
+            params.append(param)
+        else:
+            param.requires_grad = False
+    return params
+
+def set_up_optim(model, device, solver='AdamW', dp=False, finetune=-1, sample_rate=0.01,
+            alphas=[3,10,100], noise_multiplier=0.01, max_grad_norm=0.1, batch_size=1,
+            warmup_steps=5000, lr=0.001, Huggingface=False, perturb=False):
+    model =  model.to(device)
+    if Huggingface:
+        params = model.parameters()
+    else:
+        params = finetunable_GPT2_params(model,finetune)
+    if solver == 'AdamW':
+        opt = AdamW
+    elif solver == 'SGD':
+        opt = optim.SGD
+    optimizer = opt(params, lr=lr)
+    if dp:
+        model.train()
+        privacy_engine = PrivacyEngine(
+            model,
+            alphas=alphas,
+            sample_rate=sample_rate,
+            noise_multiplier=noise_multiplier,
+            max_grad_norm=max_grad_norm,
+            batch_size=batch_size
+        )
```

Comment:
Something is off here
As far as I remember, `PrivacyEngine` should throw an exception if you provide both `batch_size` and `sample_rate` at the same time - if it doesn't it's a bug.

If you want a long-is story why, here it is:
- `batch_size` is a part of old API
- Then we brought the implementation closer to its theoretical foundations. Privacy accounting implemented in opacus assumes, that minibatches are formed by flipping a biased coin for each record in the dataset - and each record is included in a current batch with a certain probability q. With this approach you no longer have a fixed batch size - it'll vary somewhat each time. So we deprecated `batch_size` parameter in favour of `sample_rate`
- The two are supposed to be mutually exclusive and defined through each other: `sample_rate = batch_size / len(dataset)`